modules/mod-deep-learning/DeepLearningEffect.cpp

```diff
@@ -0,0 +1,182 @@
+
+// ProcessOne() takes a track, transforms it to bunch of buffer-blocks,
+// performs a forward pass through the deep model, and writes 
+// the output to new tracks. 
+bool DeepLearningEffect::ProcessOne(WaveTrack *leader,
+                           double tStart, double tEnd)
+{
+   if (leader == NULL)
+      return false;
+
+   // keep track of the sample format and rate,
+   // we want to convert all output tracks to this
+   sampleFormat origFmt = leader->GetSampleFormat();
+   int origRate = leader->GetRate();
+
+   // initialize source tracks, one for each source that we will separate
+   std::vector<WaveTrack::Holder> sourceTracks;
+   std::vector<std::string> sourceLabels = mModel->GetCard()->labels();
+   sourceTracks = CreateSourceTracks(leader, sourceLabels);
+   
+   // Initiate processing buffer, most likely shorter than
+   // the length of the selection being processed.
+   Floats buffer{ leader->GetMaxBlockSize() };
+
+   // get each of the blocks we will process
+   std::vector<BlockIndex> indices = GetBlockIndices(leader, tStart, tEnd);
+   for (BlockIndex block : indices)
+   {
+      //Get a blockSize of samples (smaller than the size of the buffer)
+      sampleCount samplePos = block.first;
+      size_t blockSize = block.second;
+      double tPos = leader->LongSamplesToTime(samplePos); 
+   
+      // get a torch tensor from the leader track
+      torch::Tensor input = BuildMultichannelTensor(leader, buffer.get(), 
+                                            samplePos, blockSize);
+
+      // if we're not doing a multichannel forward pass, downmix
+      if (!mModel->GetCard()->multichannel())
+         input = input.sum(0, true, torch::kFloat);
+
+      // resample!
+      input = mModel->Resample(input, origRate, mModel->GetSampleRate());
+
+      // forward pass!
+      torch::Tensor output = mModel->ToTensor(ForwardPassInThread(input));
+
+      // resample back
+      output = mModel->Resample(output, mModel->GetSampleRate(), origRate);
+
+      // write each source output to the source tracks
+      for (size_t idx = 0; idx < output.size(0) ; idx++)
+         TensorToTrack(output[idx].unsqueeze(0), sourceTracks[idx], 
+                       tPos, tEnd); 
+
+      // Update the Progress meter
+      mCurrentProgress = (tPos - tStart) / (tEnd - tStart);
+      if (TrackProgress(mCurrentTrackNum, mCurrentProgress)) 
+         return false;
+   }
+
+   // postprocess the source tracks to the user's sample rate and format
+   PostProcessSources(leader, sourceTracks, origFmt, origRate);
+
+   return true;
+}
+
+std::vector<WaveTrack::Holder> DeepLearningEffect::CreateSourceTracks
+(WaveTrack *leader, std::vector<std::string> &labels)
+{
+   std::vector<WaveTrack::Holder> sources;
+   for (auto &label : labels)
+   {
+      WaveTrack::Holder srcTrack = leader->EmptyCopy();
+
+      // append the source name to the track's name
+      srcTrack->SetName(srcTrack->GetName() + wxString("-" + label));
```

Comment:
See corresponding place in DeepLearningAnalyzer.  Maybe we need to be careful about internationalization of names of track.  Maybe it would be nice to translate the labels.  But labels depend on some imported model and are not known from the source code tree.  (They come ultimately from the .json data?) An open question how this might be fixed.

Reply:
yes, the labels do come from the .json data, meaning that they would have to be translated on the fly somehow.